hevc/slice.go

```diff
@@ -0,0 +1,485 @@
+	UseIntegerMvFlag                  bool
+	QpDelta                           int
+	CbQpOffset                        int8
+	CrQpOffset                        int8
+	ActYQpOffset                      int8
+	ActCbQpOffset                     int8
+	ActCrQpOffset                     int8
+	CuChromaQpOffsetEnabledFlag       bool
+	DeblockingFilterOverrideFlag      bool
+	DeblockingFilterDisabledFlag      bool
+	BetaOffsetDiv2                    int8
+	TcOffsetDiv2                      int8
+	LoopFilterAcrossSlicesEnabledFlag bool
+	NumEntryPointOffsets              uint
+	OffsetLenMinus1                   uint8
+	EntryPointOffsetMinus1            []uint32
+	SegmentHeaderExtensionLength      uint16
+	SegmentHeaderExtensionDataByte    []byte
+	Size                              uint32
+}
+
+type RefPicListsModification struct {
+	RefPicListModificationFlagL0 bool
+	ListEntryL0                  []uint8
+	RefPicListModificationFlagL1 bool
+	ListEntryL1                  []uint8
+}
+
+type PredWeightTable struct {
+	LumaLog2WeightDenom        uint8
+	DeltaChromaLog2WeightDenom int8
+	WeightsL0                  []Weight
+	WeightsL1                  []Weight
+}
+
+type Weight struct {
+	LumaWeightFlag    bool
+	ChromaWeightFlag  bool
+	DeltaLumaWeight   int8
+	LumaOffset        int
+	DeltaChromaWeight [2]int8
+	DeltaChromaOffset [2]int
+}
+
+func ParseSliceHeader(nalu []byte, spsMap map[uint32]*SPS, ppsMap map[uint32]*PPS) (*SliceHeader, error) {
+	sh := &SliceHeader{}
+
+	buf := bytes.NewBuffer(nalu)
+	r := bits.NewAccErrEBSPReader(buf)
+
+	naluHdrBits := r.Read(16)
+	naluType := GetNaluType(byte(naluHdrBits >> 8))
+	sh.FirstSliceSegmentInPicFlag = r.ReadFlag()
+	if naluType >= NALU_BLA_W_LP && naluType <= NALU_IRAP_VCL23 {
+		sh.NoOutputOfPriorPicsFlag = r.ReadFlag()
+	}
+	sh.PicParameterSetId = uint32(r.ReadExpGolomb())
+	pps, ok := ppsMap[sh.PicParameterSetId]
+	if !ok {
+		return nil, fmt.Errorf("pps ID %d unknown", sh.PicParameterSetId)
+	}
+	sps, ok := spsMap[pps.SeqParameterSetID]
+	if !ok {
+		return nil, fmt.Errorf("sps ID %d unknown", pps.SeqParameterSetID)
+	}
+
+	if !sh.FirstSliceSegmentInPicFlag {
+		if pps.DependentSliceSegmentsEnabledFlag {
+			sh.DependentSliceSegmentFlag = r.ReadFlag()
+		}
+		/*
+			MinCbLog2SizeY = log2_min_luma_coding_block_size_minus3 + 3
+			CtbLog2SizeY = MinCbLog2SizeY + log2_diff_max_min_luma_coding_block_size
+			CtbSizeY = 1 << CtbLog2SizeY
+		*/
+		CtbSizeY := uint(1 << (sps.Log2MinLumaCodingBlockSizeMinus3 + 3 + sps.Log2DiffMaxMinLumaCodingBlockSize))
+		/*
+			PicWidthInCtbsY = Ceil( pic_width_in_luma_samples ÷ CtbSizeY )
+			PicHeightInCtbsY = Ceil( pic_height_in_luma_samples ÷ CtbSizeY )
+			PicSizeInCtbsY = PicWidthInCtbsY * PicHeightInCtbsY
+		*/
+		PicSizeInCtbsY := ceilDiv(uint(sps.PicWidthInLumaSamples), CtbSizeY) *
+			ceilDiv(uint(sps.PicHeightInLumaSamples), CtbSizeY)
+		sh.SegmentAddress = r.Read(bits.CeilLog2(PicSizeInCtbsY))
+	}
+
+	if !sh.DependentSliceSegmentFlag {
+		/*
+				NumPicTotalCurr = 0
```

Comment:
Please add an explanatory comment on why this out-commented code is here.

Reply:
Just write that it is pseudo-code from the standard.

Reply:
Done

Reply:
👍